feat: UIG-2809 - form - validatie verbeteringen

ksvlaanderen opened the pull request:
min-en max-exclusive validatie toegevoegd
pattern validatie documentatie voorzien
masks uitgebreid
masks storybook en cypress uitgebreid
error afbeelden met kleine top-margin
story controls refactor met constante
id en name attributen refactor
webComponent decorator gebruikt

[Jira](https://www.milieuinfo.be/jira/browse/UIG-2809)
[Bamboo](https://www.milieuinfo.be/bamboo/browse/UIGOV-CUWC270)

Goldflow commented on libs/form/src/next/error-message/vl-error-message.component.ts:
```diff
@@ -1,32 +1,33 @@
+    // Attributes
+    private show = errorMessageDefaults.show;
+    private for = errorMessageDefaults.for; // Wordt enkel gebruikt in de form-control basis klasse
+    private state = errorMessageDefaults.state; // Wordt enkel gebruikt in de form-control basis klasse
```
eventueel hier ook `// State` erboven

Goldflow replied:
Zoals besproken, is dit juist zoals het nu is.

Goldflow commented on libs/form/src/next/error-message/vl-error-message.component.ts:
```diff
@@ -1,32 +1,33 @@
         return {
+            show: { type: Boolean, reflect: true },
             for: { type: String },
             state: { type: String },
```
eventueel `state: { type: String, state: true }` ?

Goldflow replied:
Mijn commentaar klopt niet; resolved bij deze.

Goldflow commented on libs/form/src/next/input-field-masked/stories/vl-input-field-masked.stories-arg.ts:
```diff
@@ -1,44 +1,41 @@
-            category: CATEGORIES.ATTRIBUTES,
-            defaultValue: { summary: inputFieldMaskedArgs.pattern },
-        },
+        if: { arg: '' }, // Dit attribuut mag niet gebruikt worden bij dit component, verberg het in Storybook.
```
Op zich als het werkt, is het belangrijkste, maar dit is eerder bedoeld om conditionele controls af te beelden.

De Storybook manier om dit te doen is:

```
CascaderDefault.parameters = {
    controls: {
        exclude: ['content', 'label', 'labelSlot'],
    },
};
```

Dan alles je dit wil toepassen voor alle stories, kan je dit doen:

```
export default {
    ...
    parameters: {
        ...
        controls: { exclude: ['data-vl-hover'] },
    },
} as Meta<typeof dataTableArgs>;
```

Goldflow commented on libs/form/src/next/datepicker/stories/vl-datepicker.stories-arg.ts:
```diff
@@ -36,19 +36,11 @@ export const datepickerArgTypes: ArgTypes<typeof datepickerArgs> = {
-        options: ['range', 'time', 'date-time', ''],
+        description: 'Type van de datepicker',
+        control: { type: CONTROLS.SELECT },
+        options: DATEPICKER_TYPES,
```
👍🏼

Goldflow commented on libs/form/src/next/input-field-masked/stories/vl-input-field-masked.stories-doc.mdx:
```diff
@@ -36,10 +34,93 @@ Bij het `price` mask wordt de prefix weggelaten.
 
 ## Validatie
```
Well done on the extensive & clear documentation!

Goldflow commented on libs/form/src/next/input-field/stories/vl-input-field.stories-doc.mdx:
```diff
@@ -29,6 +29,39 @@ import { VlInputFieldComponent } from '@domg-wc/form/next/input-field';
 
 <FormControlPublicMethods/>
 
+## Pattern validatie
```
👍🏼

Goldflow commented on libs/form/src/next/input-field/stories/vl-input-field.stories-arg.ts:
```diff
@@ -103,22 +103,52 @@ export const inputFieldArgTypes: ArgTypes<typeof inputFieldArgs> = {
         },
     },
+    minExclusive: {
+        name: 'min-exclusive',
```
Dit is voor mij zeker geen must om te veranderen, maar gezien `min` en `max` ook al een shorthand is voor `minimum` en `maximum`, zou het dan niet ok zijn om ook `min-excl` te gebruiken voor de attribuut name? (en gelijkaardig dan voor max; `max-excl`)

Ok voor mij om het zo te houden hoor.

ksvlaanderen replied:
Besproken en gekozen om `exclusive` voluit te schrijven.

Goldflow commented on libs/form/src/next/input-field-masked/stories/vl-input-field-masked.stories-doc.mdx:
```diff
@@ -36,10 +34,93 @@ Bij het `price` mask wordt de prefix weggelaten.
+- Er wordt automatisch mask validatie toegevoegd aan het input veld, je kan dit uitschakelen met het `disable-mask-validation` attribuut.
+- Bij de mask validatie wordt er gebruik gemaakt van een regex, deze kan je overschrijven met de `regex` property.
+- Bij het testen van de regex wordt altijd de raw value van het input veld gebruikt.
+- De `patternMismatch` ValidityState key wordt gebruikt voor de mask validatie error.
```
Eventueel nog link naar https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern toevoegen en/of ook expliciet vermelden dat dit `patternMismatch` ValidityState key gebruikt zal worden voor zowel `pattern` als `regex`-variant.

ksvlaanderen replied:
De input-field-masked maakt enkel gebruik van de regex, pattern kan je hier niet gebruiken.
Om verwarring te vermijden plaats ik deze link er liever niet bij.

Goldflow commented on libs/form/src/next/input-field/stories/vl-input-field.stories-doc.mdx:
```diff
@@ -29,6 +29,39 @@ import { VlInputFieldComponent } from '@domg-wc/form/next/input-field';
+
+Het `pattern` attribuut kan gebruikt worden voor eenvoudige validatie.<br/>
+De `regex` property kan gebruikt worden voor complexere validatie.
+
```
Eventueel nog toevoegen welke ValidityState key getriggerd word bij deze validatie gelijkaardig zoals je deed bij `vl-input-field-masked`.

Zouden we het verschaffen van die info ook mogelijks doortrekken voor de andere attributen die overeenkomen met specifieke ValidityState keys? die info kan ofwel eventueel bij de Storybook-`args` komen of ergens anders op een centrale plaats. Heb deze comment ook toegevoegd in toekomstige jira task ivm documentatie - kunnen dat dan nog altijd laten vallen.

ksvlaanderen replied:
Is het goed als we dit op een ander punt opnemen?
We zouden ergens meer informatie moeten aanbieden ivm de standaarden die we volgen, het gebruik van de open-wc/form-participation package en de ElementInternals API.

Goldflow replied:
Is goed. Ik had ivm ValidityState al reeds toegevoegd in [UIG-2842](https://www.milieuinfo.be/jira/browse/UIG-2842). Heb jouw comment nu hier ook nog in toegevoegd. Dan nemen we dit later op.

Goldflow reviewed (commented):
Good job!

Op technisch vlak geen enkele opmerking; ziet er goed uit. Ook de documentatie is al heel goed. Heb enkele kleinere opmerkingen toegevoegd die eventueel ook later kunnen opgenomen worden of niet.

ksvlaanderen committed:
feat: UIG-2809 - form - validatie verbeteringen

min-en max-exclusive validatie toegevoegd
pattern validatie documentatie voorzien
masks uitgebreid
masks storybook en cypress uitgebreid
error afbeelden met kleine top-margin
story controls refactor met constante
id en name attributen refactor
webComponent decorator gebruikt

Goldflow reviewed (approved):
🎸 🚀 💯

kspeltix commented:
:tada: This PR is included in version 1.27.0 :tada:

The release is available on [GitHub release](https://github.com/milieuinfo/uigov-web-components/releases/tag/v1.27.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
